plugin/plugins/filesystem/nodes/base.go

```diff
@@ -0,0 +1,275 @@
+package fsnodes
+
+import (
+	"context"
+	"errors"
+	gopath "path"
+	"time"
+
+	"github.com/djdv/p9/p9"
+	"github.com/djdv/p9/unimplfs"
+	nodeopts "github.com/ipfs/go-ipfs/plugin/plugins/filesystem/nodes/options"
+	logging "github.com/ipfs/go-log"
+	coreiface "github.com/ipfs/interface-go-ipfs-core"
+	corepath "github.com/ipfs/interface-go-ipfs-core/path"
+)
+
+const ( //device - attempts to comply with standard multicodec table
+	dMemory = 0x2f // generic "path"
+	dIPFS   = 0xe4
+)
+
+var _ p9.File = (*Base)(nil)
+
+type p9Path = uint64
+
+// Base provides a foundation to build file system nodes which contain file meta data
+// as well as stubs for unimplemented file system methods
+type Base struct {
+	// Provide stubs for unimplemented methods
+	unimplfs.NoopFile
+	p9.DefaultWalkGetAttr
+
+	Trail []string // FS "breadcrumb" trail from node's root
+
+	// Storage for file's metadata
+	Qid      *p9.QID
+	meta     *p9.Attr
+	metaMask *p9.AttrMask
```

Comment:
The intent is to share metadata across multiple references that point to the same path.
During a walk-clone, we copy the pointers to the actual structs and de-reference them when reading/writing from/to them.
In that case, if either reference modifies some metadata, it should be reflected on the other reference as well without needing to fetch it again.

When walking to a new path, we allocate for new metadata of that path.

This will cause issues when moving back and forth though (walking back to `..`), so we should probably store metadata in an index on the FS instead.
Something similar to
 ```go
func (node *FSNode) GetAttr() p9.Attr {
node.metaLock()
defer node.metaUnlock()

var meta *p9.Attr = node.metaMap[node.path]
meta.whatever = "modified"
return *meta
}
```

Reply:
At the time, I was thinking about the scenario where you have 2 references to the same path.
You then take one of them, step down a directory, then step back with `..`.

With this, we'd start with a node, that shares metadata with the other reference.
When you step down, we "move" the node by changing it's path and allocate new metadata for it.
We do the same when stepping back to `..`.
Which means the 2 nodes that originally pointed to the same attr, no longer do, and are instead completely independent.

This will cause issues if we expect to get back the exact node we came from when resolving "..", which we may need here to track `open` state properly.

\-\-
We also talked about generating metadata when needed rather than storing all of it and relying on that.
In addition, we could (and probably should) link each node together during Walk/Step, meaning a `..` would simply return `node.parent` rather than changing the current node's path, or doing any other processing.

***